pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,520 @@
+            specification of Debye-Waller factors. Note that these
+            factors are temperature dependent.
+            cs (float): the chromatic aberration coefficient. set by default to 1 mm.
+            later on: may want "number of iterations", "magnification", "critical value of beam",
+            "twin direction" for certain materials, "sample thickness", and "excitation error s"
+        """
+        self.wavelength_cache = wavelength_cache
+        self.symprec = symprec
+        self.voltage = voltage
+        self.beam_direction = beam_direction
+        self.camera_length = camera_length
+        self.debye_waller_factors = debye_waller_factors or {}
+        self.cs = cs
+
+    def wavelength_rel(self) -> float:
+        """
+        Calculates the wavelength of the electron beam with relativstic kinematic effects taken
+        into account (electrons are way faster than X-rays, so you can't neglect these effects).
+        Args:
+            none
+        Returns:
+            relativisticWavelength (in meters)
+        """
+        if self.voltage in self.wavelength_cache:
+            return self.wavelength_cache[self.voltage]
+        wavelength_rel = sc.h / np.sqrt(2 * sc.m_e * sc.e * 1000 * self.voltage *
+                                        (1 + (sc.e * 1000 * self.voltage) / (2 * sc.m_e * sc.c ** 2)))
+        self.wavelength_cache[self.voltage] = wavelength_rel
+        return wavelength_rel
+
+    def generate_points(self, coord_left: int = -10, coord_right: int = 10) -> np.ndarray:
+        """
+        Generates a bunch of 3D points that span a cube.
+        Args:
+            coord_left (int): The minimum coordinate value.
+            coord_right (int): The maximum coordinate value.
+        Returns:
+            Numpy 2d array
+        """
+        points = [0, 0, 0]
+        coord_values = np.arange(coord_left, coord_right + 1)
+        points[0], points[1], points[2] = np.meshgrid(coord_values, coord_values, coord_values)
+        points_matrix = (np.ravel(points[i]) for i in range(0, 3))
+        result = np.vstack(list(points_matrix)).transpose()
+        return result
+
+    def zone_axis_filter(self, points: np.ndarray, laue_zone: int = 0) -> List[Tuple[int, int, int]]:
+        """
+        Filters out all points that exist within the specified Laue zone according to the zone axis rule.
+        Args:
+            points (np.ndarray): The list of points to be filtered.
+            laue_zone (int): The desired Laue zone.
+        Returns:
+            list of 3-tuples
+        """
+        if any(isinstance(n, tuple) for n in points):
+            return points
+        if len(points) == 0:
+            return []
+        filtered = np.where(np.dot(np.array(self.beam_direction), np.transpose(points)) == laue_zone)
+        result = points[filtered]
+        result_tuples = [tuple(x) for x in result.tolist()]
+        return result_tuples  # type: ignore
```

Comment:
Why `# type: ignore`?

Reply:
When I run mypy, it keeps giving me an error that I have Tuple[any, any...] even though I checked that they are int inside the tuple. If I change the type hint to Tuple[any, any..] instead, it creates several more errors because other methods are specified as Tuple[int, int, int]... So to keep the type hints as meaningful as possible I added it

Reply:
Ah, yes. I suspect that's because it doesn't know the length of `result`. You may be able to fix by explicitly declaring `result_tuples: Tuple[int, int, int] = ...`, I'm not sure. You could also try `Tuple[int, ...]` maybe. Annoying!

Reply:
I will take another look today ASAP! Thanks

Reply:
Thanks, and not the end of the world if you can't fix. Type hints are definitely preferred but they are optional for non-core classes, and the method signature is the main thing the matters.

Reply:
Thank you! I tried different combinations of type hints for multiple methods to make them work... but mypy still won't recognize it. I am now casting the result to be List[Tuple[int, int, int]] (even though when I run the debugger, it is already List[Tuple[int, int, int]] without casting...).